### PR TITLE
doc/lib-functions: remove warnings, add `versions`, `cli`

### LINKED ISSUE #116472: nixpkgs manual: warning: lib.readPathsFromFile is deprecated
**Describe the bug**

I get this warnings when opening nix-shell for manual:

```
[davidak@gaming:~/code/nixpkgs/doc]$ nix-shell 
trace: lib.crossLists is deprecated, use lib.cartesianProductOfSets instead
trace: `mkStrict' is obsolete; use `mkOverride 0' instead.
trace: `lib.nixpkgsVersion` is deprecated, use `lib.version` instead!
trace: warning: lib.readPathsFromFile is deprecated, use a list instead
trace: Warning: `showVal` is deprecated and will be removed in the next release, please use `traceSeqN`
trace: lib.zip is deprecated, use lib.zipAttrsWith instead
these derivations will be built:
  /nix/store/9zz2jxwyb2abix1fy24hpdkxd3caa92j-epub.xsl.drv
  /nix/store/cd41m3dip3ygi6s2mfv066cc0jzq0x9l-xhtml.xsl.drv
  /nix/store/qcg7k535a5rs3rkf4kzz93rdcvn904s4-locations.xml.drv
  /nix/store/iwvmrda8ypzg6h4wqh90xnivmb1myf9h-nixpkgs-lib-docs.drv
  /nix/store/68v6x94cpnvnbaynch6g7ncwl7647sdv-doc-support.drv
these paths will be fetched (351.53 MiB download, 675.01 MiB unpacked)
...
```

It still works.

This was first mentioned in https://github.com/NixOS/nixpkgs/issues/63513#issuecomment-786770936

Might get fixed by https://github.com/NixOS/nixpkgs/pull/108063

**To Reproduce**
1. Follow [Contributing to this documentation](https://nixos.org/manual/nixpkgs/unstable/#chap-contributing)

**Expected behavior**
no warnings

**Additional context**

**Notify maintainers**

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.11.2, NixOS, 21.05pre274338.102eb68ceec (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"home-manager-20.09, nixos-21.05pre274338.102eb68ceec, nixos-hardware, nixos-unstable-21.05pre274338.102eb68ceec"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by ncfavier**
Building the nixpkgs manual currently triggers a bunch of deprecation warnings, because every attribute in `lib` is evaluated to see if it's an attrset to generate locations for.

Instead, share the lib subsets to include in the documentation between `lib-function-docs` and `lib-function-locations` so they can coordinate.

I also wanted to add `modules` (the [upcoming](https://github.com/NixOS/nixpkgs/pull/204840) `lib.modules.typeCheck{,ing}` might especially be worth documenting) but there are several obstacles:
- that file contains many functions that are too internal to be exposed in the docs. Maybe they should actually be made local, or split into a different file if we don't want to break the API?
- `nixdoc` generates documentation for non-toplevel local functions like [`collectStructuredModules`](https://github.com/NixOS/nixpkgs/blob/99fed31f9da9265b6b608955610e4b72247e0dad/lib/modules.nix#L362-L387) (which then fail the build because they try to include location information that wasn't generated).

Fixes https://github.com/NixOS/nixpkgs/issues/116472

### CODE REVIEW

File: doc/doc-support/lib-function-docs.nix [hunk 1]
```diff
@@ -1,12 +1,10 @@
-# Generates the documentation for library functons via nixdoc. To add
-# another library function file to this list, the include list in the
-# file `doc/functions/library.xml` must also be updated.
```
Comment (roberth) [inline doc/doc-support/lib-function-docs.nix:3]: Isn't this still the case?
I think it's quite feasible to generate that file though, in the derivation below.

Comment (ncfavier) [reply doc/doc-support/lib-function-docs.nix:3]: I moved the comment along with the list, in `doc/doc-support/default.nix`. But yes that would be nice.

File: doc/doc-support/lib-function-docs.nix [hunk 2]
```diff
@@ -1,12 +1,10 @@
...
+# Generates the documentation for library functons via nixdoc.
 
-{ pkgs ? import ./.. {}, locationsXml }:
+{ pkgs ? import ./.. {}, locationsXml, libsets }:
```
Comment (roberth) [inline doc/doc-support/lib-function-docs.nix:3]:
```suggestion
{ pkgs, locationsXml, libsets }:
```

I don't think this file will be auto-called, making the default either duplicated or dead code.

File: doc/doc-support/lib-function-locations.nix [hunk 3]
```diff
@@ -1,4 +1,4 @@
-{ pkgs ? (import ./.. { }), nixpkgs ? { }}:
+{ pkgs ? (import ./.. { }), nixpkgs ? { }, libsets }:
```
Comment (roberth) [inline doc/doc-support/lib-function-locations.nix:1]:
```suggestion
{ pkgs, nixpkgs, libsets }:
```

I don't think this file will be auto-called, making the default both duplicated and dead code.

**Review by fricklerhandwerk — APPROVED** [review]
Looks good! Please implement @roberth's suggestions.

**Comment by ncfavier** [comment]
I struggled a bit with XML but here we go.

Note that the list of sections is ordered alphabetically now, which is probably better?

I also bumped the `generate.section.toc.level` so that we get TOCs for X.Y sections.

**Comment by fricklerhandwerk** [comment]
@ncfavier good point about the ordering, that indeed eluded me. I'm skeptical about changing the order haphazardly, as it (implicitly, but still) tells readers of the TOC something about relative importance and can be used to guide their attention.

Even if it's more cumbersome I'd prefer to instead pass the sections around as a list. @roberth what do you think?

**Comment by roberth** [comment]
> it (implicitly, but still) tells readers of the TOC something about relative importance and can be used to guide their attention.

I don't think the ordering was particularly good, but I guess it works?

> relative importance

Seems a bit fanciful for independent pieces of reference documentation.

-----


Perhaps it may be helpful (for the TOC) to start the section names with `lib.${name}`, or just `${name}`:

**Comment by ncfavier** [comment]
> Perhaps it may be helpful (for the TOC) to start the section names with `lib.${name}`, or just `${name}`:

Done, but nixdoc escapes `<` to `&gt;` so can't wrap it in `<literal>`.

**Review by roberth — APPROVED** [review]

**Review by infinisil — APPROVED** [review]

**Pull request merged**